File: chapter4/quick_sort/1.go
package main

import (
	"fmt"
)

func quickSort(list []int) []int {
	length := len(list)

	if length < 2 {
		return list
	}

	pivotIndex := length / 2
	pivot := list[pivotIndex]

	newList := append(list[:pivotIndex], list[pivotIndex+1:]...)

	lesser := []int{}
	greater := []int{}

	for _, x := range newList {
		if x <= pivot {
			lesser = append(lesser, x)
		} else {
			greater = append(greater, x)
		}
	}

	result := []int{}

	result = append(result, quickSort(lesser)...)
	result = append(result, pivot)
	result = append(result, quickSort(greater)...)

	return result
}

func main() {
	fmt.Println(quickSort([]int{10, 5, 2, -15, 2, 3, -10, 7}))
}

File: chapter4/quick_sort/1.js
const quickSort = (list) => {
  const length = list.length;

  if (length < 2) {
    return list;
  }

  const pivotIndex = Math.floor(length / 2);
  const pivot = list[pivotIndex];

  const newList = [...list.slice(0, pivotIndex), ...list.slice(pivotIndex + 1)]

  const lesser = [];
  const greater = [];

  newList.forEach((x) => {
    if (x <= pivot) {
      lesser.push(x);
    } else {
      greater.push(x);
    }
  });

  return [...quickSort(lesser), pivot, ...quickSort(greater)];
};

console.log(quickSort([10, 5, 2, -15, 2, 3, -10, 7]));